### ISSUE TITLE
PyTorch Ray Estimator seems not syncing weights for different workers

### ISSUE THREAD

**Issue by hkvision** [intel-analytics/analytics-zoo#3184]
If I intentionally add `time.sleep` in `data_creator` to let different workers randomly sleep for different amount of time, the workers don't start the training together (i.e. the worker sleeps for less time starts first without waiting for other workers to sync weights).
Seems each worker is training separately and they are not syncing weights.
Also if I intentionally make each worker have different data size, the worker with larger data will continue training until the training finishes.
This behavior occurs for both torch and horovod backends.

**Comment by hkvision** [comment]
Horovod result:
```
(pid=17328) Sleeping for 5 seconds
(pid=17329) Sleeping for 1 seconds
(pid=17329) Number of files on this worker:  8
(pid=17329) Number of files to be trained:  6
  0%|          | 0/1172 [00:00<?, ?batch/s]
(pid=17329) Data size on worker:  75000
  0%|          | 1/1172 [00:00<00:47, 24.62batch/s, loss=0.735]
(pid=17329) Worker 0 trained batch 0
  0%|          | 4/1172 [00:00<00:31, 37.48batch/s, loss=0.621]
(pid=17329) Worker 0 trained batch 1
(pid=17329) Worker 0 trained batch 2
(pid=17329) Worker 0 trained batch 3
(pid=17329) Worker 0 trained batch 4
(pid=17329) Worker 0 trained batch 5
  1%|          | 6/1172 [00:00<00:30, 38.33batch/s, loss=0.506]
  1%|          | 8/1172 [00:00<00:29, 39.31batch/s, loss=0.507]
(pid=17329) Worker 0 trained batch 6
(pid=17329) Worker 0 trained batch 7
  1%|          | 10/1172 [00:00<00:29, 40.01batch/s, loss=0.44]
(pid=17329) Worker 0 trained batch 8
(pid=17329) Worker 0 trained batch 9
(pid=17329) Worker 0 trained batch 10
(pid=17329) Worker 0 trained batch 11
  1%|          | 12/1172 [00:00<00:28, 40.83batch/s, loss=0.67] 
  1%|▏         | 15/1172 [00:00<00:27, 41.44batch/s, loss=0.537]
 Worker 0 trained batch 12
(pid=17329) Worker 0 trained batch 13
(pid=17329) Worker 0 trained batch 14
```
At the very beginning only worker with pid 17329 is training.
```
(pid=17328) Worker 0 trained batch 17
(pid=17329) Worker 0 trained batch 193
(pid=17328) Worker 0 trained batch 18
(pid=17329) Worker 0 trained batch 194
 17%|█▋        | 195/1172 [00:05<00:25, 38.19batch/s, loss=0.534]
  2%|▏         | 19/782 [00:01<00:41, 18.47batch/s, loss=0.534]
(pid=17328) Worker 0 trained batch 19
(pid=17329) Worker 0 trained batch 195
(pid=17329) Worker 0 trained batch 196
 17%|█▋        | 197/1172 [00:05<00:25, 37.87batch/s, loss=0.51] 
  3%|▎         | 20/782 [00:01<00:41, 18.47batch/s, loss=0.557]
(pid=17328) Worker 0 trained batch 20
(pid=17329) Worker 0 trained batch 197
  3%|▎         | 21/782 [00:01<00:40, 18.56batch/s, loss=0.561]
 17%|█▋        | 199/1172 [00:05<00:25, 37.50batch/s, loss=0.569]
  3%|▎         | 22/782 [00:01<00:40, 18.59batch/s, loss=0.441]
(pid=17328) Worker 0 trained batch 21
```
Afterwards, 17328 joins and they have different training progress. After 17329 finishes its training, 17328 continues.
```
100%|█████████▉| 780/782 [00:41<00:00, 18.98batch/s, loss=0.507]
(pid=17328) Worker 0 trained batch 780
(pid=17328) Worker 0 trained batch 781
(pid=17329) Worker 0 trained batch 957
 82%|████████▏ | 958/1172 [00:45<00:10, 21.17batch/s, loss=0.436]
                                                                
 82%|████████▏ | 960/1172 [00:45<00:10, 21.18batch/s, loss=0.458]
(pid=17329) Worker 0 trained batch 958
(pid=17329) Worker 0 trained batch 959
(pid=17329) Worker 0 trained batch 960
(pid=17329) Worker 0 trained batch 961
(pid=17329) Worker 0 trained batch 962
 82%|████████▏ | 963/1172 [00:45<00:09, 21.22batch/s, loss=0.477]
(pid=17329) Worker 0 trained batch 963
(pid=17329) Worker 0 trained batch 964
 82%|████████▏ | 965/1172 [00:45<00:09, 21.24batch/s, loss=0.497]
(pid=17329) Worker 0 trained batch 965
(pid=17329) Worker 0 trained batch 966
```

**Comment by hkvision** [comment]
Torch distributed result:
```
(pid=9122) Sleeping for 5 seconds
(pid=9119) Sleeping for 10 seconds
(pid=9122) Worker 1 trained batch 0
(pid=9122) Worker 1 trained batch 1
(pid=9122) Worker 1 trained batch 2
(pid=9122) Worker 1 trained batch 3
(pid=9122) Worker 1 trained batch 4
(pid=9122) Worker 1 trained batch 5
(pid=9122) Worker 1 trained batch 6
(pid=9122) Worker 1 trained batch 7
(pid=9122) Worker 1 trained batch 8
(pid=9122) Worker 1 trained batch 9
(pid=9122) Worker 1 trained batch 10

.....

(pid=9122) Worker 1 trained batch 250
(pid=9119) Worker 0 trained batch 0
(pid=9119) Worker 0 trained batch 1
  0%|          | 2/586 [00:00<00:34, 17.05batch/s, loss=0.656]
  1%|          | 4/586 [00:00<00:26, 21.73batch/s, loss=0.574]
(pid=9122) Worker 1 trained batch 251
(pid=9119) Worker 0 trained batch 2
(pid=9119) Worker 0 trained batch 3
  1%|          | 5/586 [00:00<00:26, 21.91batch/s, loss=0.597]
(pid=9122) Worker 1 trained batch 252
(pid=9119) Worker 0 trained batch 4
(pid=9122) Worker 1 trained batch 253
(pid=9119) Worker 0 trained batch 5

......

(pid=9122) Worker 1 trained batch 585
(pid=9119) Worker 0 trained batch 336
 58%|█████▊    | 337/586 [00:15<00:11, 21.72batch/s, loss=0.531]
 58%|█████▊    | 340/586 [00:15<00:11, 21.80batch/s, loss=0.518]
(pid=9119) Worker 0 trained batch 337
(pid=9119) Worker 0 trained batch 338
(pid=9119) Worker 0 trained batch 339
(pid=9119) Worker 0 trained batch 340
(pid=9119) Worker 0 trained batch 341
 58%|█████▊    | 342/586 [00:15<00:11, 21.87batch/s, loss=0.468]
(pid=9119) Worker 0 trained batch 342
(pid=9119) Worker 0 trained batch 343
 59%|█████▊    | 344/586 [00:15<00:11, 21.94batch/s, loss=0.503]
(pid=9119) Worker 0 trained batch 344
(pid=9119) Worker 0 trained batch 345
(pid=9119) Worker 0 trained batch 346
 59%|█████▉    | 347/586 [00:15<00:10, 22.05batch/s, loss=0.446]
 60%|█████▉    | 350/586 [00:15<00:10, 22.15batch/s, loss=0.382]
(pid=9119) Worker 0 trained batch 347
(pid=9119) Worker 0 trained batch 348
(pid=9119) Worker 0 trained batch 349
 60%|██████    | 352/586 [00:15<00:10, 22.22batch/s, loss=0.377]
```

**Comment by hkvision** [comment]
Torch distributed fixed in #3185 and backported.
